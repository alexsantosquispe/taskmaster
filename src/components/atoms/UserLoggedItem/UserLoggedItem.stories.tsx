import type { Meta, StoryObj } from '@storybook/react';

import UserLoggedItem from './UserLoggedItem';

const userMockItem = {
  name: 'Jhon Doe',
  email: 'jhon456@gmail.com',
  avatarUrl: '/img/avatar.webp'
};

const meta = {
  title: 'Atoms/UserLoggedItem',
  component: UserLoggedItem,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof UserLoggedItem>;

export default meta;

type Story = StoryObj<typeof UserLoggedItem>;

export const Default: Story = () => {
  return <UserLoggedItem {...userMockItem} />;
};

Default.args = {};
