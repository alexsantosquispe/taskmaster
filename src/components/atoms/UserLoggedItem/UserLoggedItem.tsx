import { Avatar } from '../Avatar/Avatar';
import { ChevronUpIcon } from '@/icons';
import { Link } from 'react-router-dom';
import cn from 'clsx';
import { currentUser } from '@/constants';
import { memo } from 'react';
import { twMerge } from 'tailwind-merge';
import { useAuth } from '@/hooks/useAuth';

interface UserLoggedItemProps {
  isCollapsed?: boolean;
}

const UserLoggedItem = ({ isCollapsed = false }: UserLoggedItemProps) => {
  const { user } = useAuth();

  return (
    <Link to="/home/profile">
      <div
        className={twMerge(
          'flex items-center rounded-lg py-2 hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/10',
          cn({
            'justify-between px-2': !isCollapsed,
            'justify-center': isCollapsed
          })
        )}
      >
        <div className="flex items-center gap-2">
          <Avatar />
          {!isCollapsed && (
            <div className="flex flex-col justify-center text-xs">
              <span>{currentUser.name}</span>
              <span className="text-neutral-600 dark:text-white/60">
                {user?.email}
              </span>
            </div>
          )}
        </div>

        {!isCollapsed && <ChevronUpIcon className="size-4 rotate-90" />}
      </div>
    </Link>
  );
};

export default memo(UserLoggedItem);
