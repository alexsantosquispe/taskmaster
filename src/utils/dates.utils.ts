import { differenceInSeconds, formatDistanceToNowStrict } from 'date-fns';

export const formatDateToDayMonth = (date: string) => {
  const options: Intl.DateTimeFormatOptions = {
    day: 'numeric',
    month: 'short'
  };
  return new Intl.DateTimeFormat('en-US', options).format(new Date(date));
};

export const formatToFriendlyDate = (date: string): string => {
  const parsedDate = new Date(date);

  if (isNaN(parsedDate.getTime())) return 'Invalid date';

  const diffSeconds = differenceInSeconds(new Date(), parsedDate);

  if (diffSeconds < 60) return 'just now';

  return `${formatDistanceToNowStrict(parsedDate)} ago`;
};
