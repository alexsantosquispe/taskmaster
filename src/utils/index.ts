import clsx from 'clsx';
import { PRIORITY_COLORS } from '../constants';
import { type PriorityType } from '../models/types';

export const formatDateToDayMonth = (date: string) => {
  const options: Intl.DateTimeFormatOptions = {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  };
  return new Intl.DateTimeFormat('en-US', options).format(new Date(date));
};

export const getPercentageStyles = (percentage: number) => {
  return clsx({
    'stroke-green-600 dark:stroke-green-500 text-green-600 dark:text-green-500':
      percentage >= 70,
    'stroke-green-500 dark:stroke-green-400 text-green-500 dark:text-green-400':
      percentage >= 56 && percentage < 70,
    'stroke-amber-500 dark:stroke-amber-400 text-amber-500 dark:text-amber-400':
      percentage >= 1 && percentage < 56,
    'stroke-primary dark:stroke-white text-neutral-600 dark:text-white/80':
      percentage === 0
  });
};

export const getPriorityStyles = (priority: PriorityType) => {
  return PRIORITY_COLORS[priority];
};
