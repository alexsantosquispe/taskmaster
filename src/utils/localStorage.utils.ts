import type { LSValuesTypes } from '../models/types';

export const getLSValue = (key: string) => localStorage.getItem(key);

export const setLSValue = (key: string, value: LSValuesTypes) => {
  localStorage.setItem(key, String(value));
};
