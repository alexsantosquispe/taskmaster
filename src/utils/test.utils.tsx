import { ToastStack } from '@/components/atoms/ToastStack/ToastStack';
import { ThemeProvider } from '@/contexts/ThemeProvider';
import { ToastProvider } from '@/contexts/ToastProvider';
import { apiClient } from '@/services/api';
import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { type ReactElement, type ReactNode } from 'react';
import { useForm, type Control, type FieldValues } from 'react-hook-form';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { twMerge } from 'tailwind-merge';

interface RenderWithFormOptions {
  defaultValues: FieldValues;
  componentToRender: (control: Control<FieldValues>) => ReactElement;
}

export const renderWithForm = ({
  defaultValues,
  componentToRender
}: RenderWithFormOptions) => {
  const Wrapper = () => {
    const { control } = useForm<FieldValues>({ defaultValues });
    return componentToRender(control);
  };

  return render(<Wrapper />);
};

interface WrapperProps {
  children: ReactNode;
  className?: string;
}

export const WrapperUI = ({ children, className }: WrapperProps) => {
  return (
    <div
      className={twMerge(
        'text-primary flex items-center justify-center rounded border border-neutral-400 p-8 dark:text-white',
        className
      )}
    >
      {children}
    </div>
  );
};

export const TestWrapper = ({ children }: WrapperProps) => {
  return (
    <ThemeProvider>
      <BrowserRouter
        future={{
          v7_startTransition: true,
          v7_relativeSplatPath: true
        }}
      >
        {children}
      </BrowserRouter>
    </ThemeProvider>
  );
};

export const ToastProviderWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <ToastProvider>
      <ToastStack />
      {children}
    </ToastProvider>
  );
};

export const ReduxWrapper = ({ children }: { children: ReactNode }) => {
  const store = configureStore({
    reducer: {
      [apiClient.reducerPath]: apiClient.reducer
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(apiClient.middleware)
  });

  return <Provider store={store}>{children}</Provider>;
};
