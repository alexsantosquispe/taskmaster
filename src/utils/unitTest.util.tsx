import { BrowserRouter } from 'react-router-dom';
import type { ReactNode } from 'react';
import { ThemeProvider } from '../contexts/ThemeProvider';

export const TestWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <ThemeProvider>
      <BrowserRouter
        future={{
          v7_startTransition: true,
          v7_relativeSplatPath: true
        }}
      >
        {children}
      </BrowserRouter>
    </ThemeProvider>
  );
};
