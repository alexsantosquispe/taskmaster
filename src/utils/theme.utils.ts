import { THEME_KEY } from '../constants';
import { Theme, type ThemeType } from '../contexts/ThemeContext';
import { getLSValue } from './localStorage.utils';

export const getSystemTheme = () => {
  if (typeof window === 'undefined') return Theme.light;

  return window.matchMedia('(prefers-color-scheme: dark)').matches
    ? Theme.dark
    : Theme.light;
};

export const getStoredTheme = (): ThemeType | null => {
  const storedTheme = getLSValue(THEME_KEY);

  if (!storedTheme) return null;

  return storedTheme as ThemeType;
};

export const applyThemeToDocument = (theme: ThemeType) => {
  const root = window.document.documentElement;

  if (theme === Theme.dark) {
    root.classList.add(Theme.dark);
  } else {
    root.classList.remove(Theme.dark);
  }
};
