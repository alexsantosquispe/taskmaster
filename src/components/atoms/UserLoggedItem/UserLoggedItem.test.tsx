import { ReduxWrapper, TestWrapper } from '@/utils/test.utils';

import { render } from '@testing-library/react';
import UserLoggedItem from './UserLoggedItem';

describe('UserLoggedItem', () => {
  it('should render the component correctly', () => {
    const component = render(
      <TestWrapper>
        <ReduxWrapper>
          <UserLoggedItem />
        </ReduxWrapper>
      </TestWrapper>
    );

    expect(component).toMatchSnapshot();
  });
});
