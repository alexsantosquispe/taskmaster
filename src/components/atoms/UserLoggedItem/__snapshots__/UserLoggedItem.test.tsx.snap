// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserLoggedItem should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/home/profile"
      >
        <div
          class="flex items-center rounded-lg py-2 hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/10 justify-between px-2"
        >
          <div
            class="flex items-center gap-2"
          >
            <div
              class="flex h-6 w-6 items-center justify-center rounded-full bg-neutral-200 text-neutral-600 dark:bg-white/20 dark:text-white"
            >
              <svg
                class="size-5"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="8"
                  r="5"
                />
                <path
                  d="M20 21a8 8 0 0 0-16 0"
                />
              </svg>
            </div>
            <div
              class="flex flex-col justify-center text-xs"
            >
              <span>
                Alex Santos
              </span>
              <span
                class="text-neutral-600 dark:text-white/60"
              />
            </div>
          </div>
          <svg
            class="size-4 rotate-90"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m18 15-6-6-6 6"
            />
          </svg>
        </div>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      href="/home/profile"
    >
      <div
        class="flex items-center rounded-lg py-2 hover:cursor-pointer hover:bg-neutral-100 dark:hover:bg-white/10 justify-between px-2"
      >
        <div
          class="flex items-center gap-2"
        >
          <div
            class="flex h-6 w-6 items-center justify-center rounded-full bg-neutral-200 text-neutral-600 dark:bg-white/20 dark:text-white"
          >
            <svg
              class="size-5"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="8"
                r="5"
              />
              <path
                d="M20 21a8 8 0 0 0-16 0"
              />
            </svg>
          </div>
          <div
            class="flex flex-col justify-center text-xs"
          >
            <span>
              Alex Santos
            </span>
            <span
              class="text-neutral-600 dark:text-white/60"
            />
          </div>
        </div>
        <svg
          class="size-4 rotate-90"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m18 15-6-6-6 6"
          />
        </svg>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
